lnwallet/channel.go

```diff
@@ -6916,16 +6913,75 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
 	// Ensure the fee rate doesn't dip below the fee floor.
 	_, weight := lc.availableBalance()
 	maxFeeRate := maxFee / (float64(weight) / 1000)
-	feeRate := chainfee.SatPerKWeight(
+	return chainfee.SatPerKWeight(
 		math.Max(maxFeeRate, float64(chainfee.FeePerKwFloor)),
 	)
+}
+
+// IdealCommitFeeRate uses the current network fee, the minimum relay fee,
+// maximum fee allocation and anchor channel commitment fee rate to determine
+// the ideal fee to be used for the commitments of the channel.
+func (lc *LightningChannel) IdealCommitFeeRate(netFee, minRelayFee,
+	maxAnchorCommitFeeRate chainfee.SatPerKWeight,
+	maxFeeAlloc float64) chainfee.SatPerKWeight {
+
+	maxFee := lc.MaxFeeRate(maxFeeAlloc)
+
+	var commitFee chainfee.SatPerKWeight
+
+	// If the channel has anchor outputs then cap the fee rate at the
+	// max anchor fee rate if that maximum is less than our max fee rate.
+	// Otherwise, cap the fee rate at the max fee rate.
+	switch lc.channelState.ChanType.HasAnchors() &&
+		maxFee > maxAnchorCommitFeeRate {
+
+	case true:
+		commitFee = chainfee.SatPerKWeight(
+			math.Min(
+				float64(netFee),
+				float64(maxAnchorCommitFeeRate),
+			),
+		)
 
-	// Cap anchor fee rates.
-	if lc.channelState.ChanType.HasAnchors() && feeRate > maxAnchorFeeRate {
-		return maxAnchorFeeRate
+	case false:
+		commitFee = chainfee.SatPerKWeight(
+			math.Min(float64(netFee), float64(maxFee)),
+		)
+	}
+
+	if commitFee >= minRelayFee {
+		return commitFee
+	}
+
+	// The commitment fee rate is below the minimum relay fee rate.
+	// If the min relay fee rate is still below the maximum fee, then use
+	// the minimum relay fee rate.
+	if minRelayFee <= maxFee {
+		return minRelayFee
 	}
 
-	return feeRate
+	// The minimum relay fee rate is more than the ideal maximum fee rate.
+	// Check if it is smaller than the absolute maximum fee rate we can
+	// use. If it is, then we use the minimum relay fee rate and we log a
+	// warning to indicate that the max channel fee allocation option was
+	// ignored.
+	absoluteMaxFee := lc.MaxFeeRate(1)
+	if minRelayFee <= absoluteMaxFee {
+		lc.log.Warn("Ignoring max channel fee allocation to " +
+			"ensure that the commitment fee is above the " +
+			"minimum relay fee.")
+
+		return minRelayFee
+	}
+
+	// The absolute maximum fee rate we can pay is below the minimum
+	// relay fee rate. The commitment tx will not be able to propagate.
+	// To give the transaction the best chance, we use the absolute
+	// maximum fee we have available and we log an error.
+	lc.log.Error("The commitment fee rate is below the current " +
+		"minimum relay fee rate.")
+
+	return absoluteMaxFee
```

Comment:
Ok i think this should be addressed now in the first 2 commits. They ensure that MaxFeeRate takes the reserve into account. I also think I found a bug in `validateFeeRate` that should be fixed now.

Reply:
I think this possibility of "use all the users fees for the commitment transaction", is kinda scary, but then again it shouldn't really ever happen since it's unlikely that the min relay fee ever gets this high. 

One thing we can do to soften the blow here is to attempt to use the smallest max fee possible here. So instead of start at 100%, we increase in an iterative manner from the current config value, up to the max fee if that gets us above the fee rate.

Reply:
> One thing we can do to soften the blow here is to attempt to use the smallest max fee possible here

Is this referring to just line 6982 or also to check at 6967? If the latter, then does that mean that even if we can meet the min relay fee initially if we were to use maxAlloc=1 we shouldn't? If the former, then do you mean something like we just take the last used commitment fee and bump by 10% or something until we reach absolute max?

Reply:
> then does that mean that even if we can meet the min relay fee initially if we were to use maxAlloc=1 we shouldn't?

I guess I'm referring more so between finding the sweet spot between the minrelay fee and the absolute maximum fee. As a basic example, let's say the max fee is 20 and the min relay fee is 5 (larger gap just to illustrate the point). With this logic we'd just to 20 rather than picking say 6.

Reply:
hmm in the example you give, this logic will return 5. The absolute max is only ever returned iff the absolute max is strictly smaller than the min relay fee